File: 01-data-ingestion-real-time/solution-proposal/scala/src/main/scala/com/becomedataengineer/exercise1/part1/api/KafkaDataProducer.scala
package com.becomedataengineer.exercise1.part1.api

import org.apache.commons.io.FileUtils
import org.apache.kafka.clients.producer.{Callback, KafkaProducer, ProducerRecord, RecordMetadata}

import java.io.File
import java.util.Properties
import java.util.concurrent.TimeUnit
import scala.collection.mutable

class KafkaDataProducer {

  // We assume here the simplicity; on production we could have a pool of available producers matching the
  // concurrency level of our app to avoid recreating them each time, or a single instance with a better
  // batching configuration
  private val KafkaProducerInstance = {
    val producerProperties = new Properties;
    producerProperties.setProperty("bootstrap.servers", "localhost:29092");
    producerProperties.setProperty("key.serializer", "org.apache.kafka.common.serialization.StringSerializer");
    producerProperties.setProperty("value.serializer", "org.apache.kafka.common.serialization.StringSerializer");
    producerProperties.setProperty("linger.ms", TimeUnit.SECONDS.toMillis(1L).toString)
    producerProperties.setProperty("batch.size", "12KB")
    new KafkaProducer[String, String](producerProperties)
  }

  private val OutputTopic = "raw_data"

  private val kafkaProducerCallbacksHolder = new KafkaProducerCallbacksHolder()

  def produceDataToKafka(visitsToIngest: VisitsToIngest) = {
    visitsToIngest.visits.foreach(visit => {
      val eventJson = JsonMapper.Instance.writeValueAsString(visit)
      KafkaProducerInstance.send(new ProducerRecord[String, String](
        OutputTopic, visit.visitId.toString, eventJson
      ), kafkaProducerCallbacksHolder.registerCallbackForMessage(eventJson))
    })

    KafkaProducerInstance.flush()
  }

  def saveUndeliveredEvents() = {
    if (kafkaProducerCallbacksHolder.eventsToDeliver.nonEmpty) {
      FileUtils.writeStringToFile(new File(s"/tmp/bde-3/homework/p1/e1/fallback_${System.currentTimeMillis()}.json"),
        kafkaProducerCallbacksHolder.eventsToDeliver.mkString("\n"), "UTF-8")
    }
  }

}

class KafkaProducerCallbacksHolder {

  val eventsToDeliver = new mutable.HashSet[String]()

  class KafkaProducerCallback(event: String) extends Callback {

    // We keep the events management in a single place, only within this class
    eventsToDeliver.add(event)

    override def onCompletion(metadata: RecordMetadata, exception: Exception): Unit = {
      if (exception == null) {
        eventsToDeliver.remove(event)
      } else {
        exception.printStackTrace()
      }
    }
  }

  def registerCallbackForMessage(event: String): Callback = {
    new KafkaProducerCallback(event)
  }

}

File: 01-data-ingestion-real-time/solution-proposal/python/exercise1/part1/api/kafka_data_producer.py
import json
import os
import pathlib
import time

from confluent_kafka import Producer, Message


class KafkaDataProducerCallbacksHolder:

    def __init__(self):
        self.events_to_deliver = []

    def register_event_to_deliver(self, event_payload):
        self.events_to_deliver.append(event_payload)

        def handle_delivery_result(error, result: Message):
            if error:
                print('Failure' + error)
            else:
                print('success')
                self.events_to_deliver.remove(event_payload)

        return handle_delivery_result


class KafkaDataProducer:

    OUTPUT_TOPIC = 'raw_data'

    def __init__(self):
        self.producer = Producer({'bootstrap.servers': 'localhost:29092',
                                  'linger.ms': 1 * 1000,  # 10 seconds
                                  'batch.size': '12KB',
                                  'retries': 3
                                  })
        self.kafka_data_producer_callbacks_holder = KafkaDataProducerCallbacksHolder()

    def produce_data_to_kafka(self, visits_to_ingest):
        for visit_json in visits_to_ingest:
            visit = json.loads(visit_json)
            self.producer.produce(
                topic=KafkaDataProducer.OUTPUT_TOPIC,
                value=bytes(visit_json, encoding='utf-8'),
                key=bytes(str(visit['visitId']), encoding='utf-8'),
                on_delivery=self.kafka_data_producer_callbacks_holder.register_event_to_deliver(visit_json)
            )

        self.producer.flush(timeout=20)

    def save_undelivered_records(self):
        if self.kafka_data_producer_callbacks_holder.events_to_deliver:
            dir_name = '/tmp/bde-3/homework/p1/e1/'
            file_name = f'fallback_{round(time.time() * 1000)}.json'
            os.makedirs(dir_name, exist_ok=True)
            pathlib.Path(f'{dir_name}/{file_name}') \
                .write_text(
                '\n'.join([event for event in self.kafka_data_producer_callbacks_holder.events_to_deliver]))
